### ISSUE TITLE
Ocarina Of Time not starting

### ISSUE THREAD

**Issue by JunielKatarn** [project64/project64#849]
Some recent change makes Ocarina Of Time (and possibly other ROMs) fail to launch.
![image](https://cloud.githubusercontent.com/assets/4507319/11993907/290e040a-a9fc-11e5-9c4a-560607f59a3b.png)

This is likely caused by one of the more recent commits.
Current context:
http://hyvart.com/download/project64/ci/0.0.0.366/Win32.zip (built from commit 341efe227d7eef0162112fd76de530d2c66ecbb2) - WORKS
http://hyvart.com/download/project64/ci/0.0.0.368/Win32.zip (built from commit 05602d1939d18563273a8e4f1d04f0f0b70393f9) - FAILS

[UPDATE]
Current workaround: Disable the debugger (Options/Advanced).

**Comment by vgturtle127** [comment]
@JunielKatarn What ROM are you using? What is the MD5?

**Comment by JunielKatarn** [comment]
![image](https://cloud.githubusercontent.com/assets/4507319/12001766/71a605be-aab5-11e5-9dec-7edf00389ace.png)

**Comment by vgturtle127** [comment]
The game works fine on my end. You are using the verified US 1.0 ROM, so the problem isn't the ROM.

What plugins are you using to launch it?

**Comment by JunielKatarn** [comment]
I'm using the "default" plugins, AFAIK:
![image](https://cloud.githubusercontent.com/assets/4507319/12014932/6f7b1922-acfa-11e5-9bbc-09b0dc47bddc.png)

Moreover, did you try the binaries I published? (specifically, http://hyvart.com/download/project64/ci/0.0.0.368/Win32.zip).
Are you able to launch OOT using that build?

Thanks :)

**Comment by JunielKatarn** [comment]
@vgturtle127 It works on your end probably because you don't have the debugger enabled (try Options/Advanced/Enable debugger).
This line throws the error:
https://github.com/project64/project64/blob/master/Source/Project64-core/N64System/Mips/MemoryVirtualMem.cpp#L2508

Context:
Previous versions of method `CMipsMemoryVM::LW_NonMemory` did not raise a break point on the default case after masking the input `PAddr` (does that mean physical address?). What they did is the following:
```
    default:
        *Value = PAddr & 0xFFFF;
        *Value = (*Value << 16) | *Value;
        return false;
        break;
```

The current code does this:
```
        switch (PAddr & 0xFFF00000)
        {
        case 0x03F00000: Load32RDRAMRegisters(); break;
        case 0x04000000: Load32SPRegisters(); break;
        case 0x04100000: Load32DPCommand(); break;
        case 0x04300000: Load32MIPSInterface(); break;
        case 0x04400000: Load32VideoInterface(); break;
        case 0x04500000: Load32AudioInterface(); break;
        case 0x04600000: Load32PeripheralInterface(); break;
        case 0x04700000: Load32RDRAMInterface(); break;
        case 0x04800000: Load32SerialInterface(); break;
        case 0x05000000: Load32CartridgeDomain2Address1(); break;
        case 0x08000000: Load32CartridgeDomain2Address2(); break;
        case 0x1FC00000: Load32PifRam(); break;
        default:
            if (bHaveDebugger())
            {
                g_Notify->BreakPoint(__FILE__, __LINE__);
            }
            m_MemLookupValue.UW[0] = PAddr & 0xFFFF;
            m_MemLookupValue.UW[0] = (m_MemLookupValue.UW[0] << 16) | m_MemLookupValue.UW[0];
        }
```
Meaning, it raises an error and (seemingly) fails to load the ROM if `PAddr & 0xFFF00000` is not within the expected bounds.

During startup, OOT provides this value for `PAddr`: `0x6000000`, which falls into the default case.
I guess, if we don't need to do any `Load32` operations for this specific value, we should at least have a do-nothing case for it.

@project64, any ideas? Should we add a case for `0x6000000`? Moreover, should break points make ROM loading fail altogether?

For what it's worth, here is the sequence of PAddr values for OOT, up to the offending address:
```
[0]		74448908
[1]		70254596
[2]		66060300
[3]		66060300
[4]		66060300
[5]		66060300
[6]		66060300
[7]		66060300
[8]		66060300
[9]		66060300
[10]	74448912
[11]	74448908
[12]	70254596
[13]	66060300
[14]	66060300
[15]	66060300
[16]	66060300
[17]	66060300
[18]	66060300
[19]	66060300
[20]	66060300
[21]	74448912
[22]	268435464
[23]	268435464
[24]	73400336
[25]	73400336
[26]	268435464
[27]	73400336
[28]	268435464
[29]	268435472
[30]	268435476
[31]	268435472
[32]	268435476
[33]	268435464
[34]	268435464
[35]	75497496
[36]	532678652
[37]	532678652
[38]	75497496
[39]	73400340
[40]	73400344
[41]	73400348
[42]	73400352
[43]	73400356
[44]	73400360
[45]	73400364
[46]	73400368
[47]	73400336
[48]	268435456
[49]	73400340
[50]	73400348
[51]	73400352
[52]	73400344
[53]	268435456
[54]	73400336
[55]	100663296
```

**Comment by cxd4** [comment]
It doesn't belong under a new, "do-nothing" case, but it does represent unsupported peripherals.
```c
#define PI_DOM1_ADDR1           0x06000000      /* to 0x07FFFFFF */
#define PI_DOM1_ADDR2           0x10000000      /* to 0x1FBFFFFF */
#define PI_DOM2_ADDR1           0x05000000      /* to 0x05FFFFFF */
#define PI_DOM2_ADDR2           0x08000000      /* to 0x0FFFFFFF */
```
Though it hardly matters because that beta stuff is unused when playing OOT anyway on the actual, original H/W.  I imagine zilmar would sooner disable the breakpoint or undo his adding it in, at least for this game.

**Comment by project64** [comment]
I am in the process of re-writing the memory handling even more dramatically at the moment which will probably cause more issues like this one. But say in zelda's case I will add a case for the addresses that @cxd4 mentioned that will not trigger a breakpoint. As I do more testing and have the unknown addresses better handled then I will likely remove the break point

**Comment by vgturtle127** [comment]
@JunielKatarn I have the debugger always enabled, and I was using the latest compiled version of Project64.

When are those builds updated? I didn't use them because I don't know what version they are, what commit they were compiled from, etc.

**Comment by JunielKatarn** [comment]
> http://hyvart.com/download/project64/ci/0.0.0.366/Win32.zip (built from commit 341efe2) - WORKS
http://hyvart.com/download/project64/ci/0.0.0.368/Win32.zip (built from commit 05602d1) - FAILS

@vgturtle127 The commit IDs are right there, next to each download link.
Also, I tested it in the Alpha builds from @project64. The issue shows up consistently.
The explanation is clear. From recent changes to `CMipsMemoryVM::LW_NonMemory`, address values like `0x6000000` now raise a run-time breakpoint (with debugger enabled).

**Comment by vgturtle127** [comment]
@JunielKatarn OK, sorry. The version numbers 0.0.0.36x were confusing me. :confused: 

It still works fine for me though.

When I tried it with the links you provided, then it fails.  :confounded: 

If I launch it using any vanilla build it runs fine, but the x64 build on your end fails to load the ROM with that error message.

That is just what is happening on my machine, anyway. :fearful:

**Comment by JunielKatarn** [comment]
> The version numbers 0.0.0.36x were confusing me

Yeah, I use those numbers to track my build bot versions. They are in no way related to official build numbers.
The root cause has been found, anyways. For the time being, I can just turn off the debugger whenever I want to play OOT.

**Comment by vgturtle127** [comment]
@JunielKatarn When you say the "root cause" you mean the changes to CMipsMemoryVM::LW_NonMemory, right?

**Comment by project64** [comment]
@vgturtle127  I started the change of reading registers in to function reading, if it was unknown location i added a breakpoint if you had debugger enabled. Zelda was reading an unhandled location so was thowing a breakpoint

**Comment by cxd4** [comment]
Since Zelda's usage of that particular part of the RCP memory map is more of a dummy beta thing than anything required (could be 64DD-related even, not sure), it's probably fine to just dodge the breakpoint if Project64 can detect the game being played is OOT specifically (header[0x3C], header[0x3D] == 'ZL') and keep the breakpoint for other discoveries yet to be found so zilmar will have more to test against when implementing that new support from RCP memory.

**Comment by cxd4** [comment]
can't push any commits tho because no Internet for me outside of scavenger hunt discoveries, and my Internet is Linux only Windows won't detect networks on my machine anymore

**Comment by vgturtle127** [comment]
@cxd4 What happened to your internet/machine? Did you get a new wireless adapter or something?

@project64 OK, I see. I assumed some more cleanup and re-organization was probably to blame. :smile:

**Comment by cxd4** [comment]
nope, just moved and not going to pay for it

**Comment by vgturtle127** [comment]
@cxd4 So you aren't going to pay for internet? How do you get online now?

**Comment by PFCKrutonium** [comment]
I imagine probably neighbors/cell phone internet.

**Comment by vgturtle127** [comment]
Well, he said his internet was "Linux only" so I assumed it had something to do with hardware...

**Comment by PFCKrutonium** [comment]
That too, but he also said he won't be paying for it, so I took the stealing service angle... I suppose there is only one way to know for sure though.

**Comment by cxd4** [comment]
No you're right.  vgturtle just doesn't always read what I say the way I assume everyone normally should.

It's both.  My internet used to work on Windows however...something about the device software must have just died overtime.  It only connects when I boot into Linux now.

**Comment by vgturtle127** [comment]
@cxd4 Huh, I have never heard of anything like that. I have heard of people having trouble with internet in Linux, but not Windows. :laughing: 

Maybe uninstall the driver in Windows and re-install it?

**Comment by cxd4** [comment]
I would need the Internet to be able to reinstall it on Windows...conveniently.

Trouble with internet on Linux:  Yes I've had it, when booting into Slackware, or other distributions that default to not supplying any wifi drivers, which are proprietary non-free and closed-source.

**Comment by vgturtle127** [comment]
I meant download them through Linux and then put them on a flash drive or something. :laughing:

**Comment by cxd4** [comment]
tl;dr internet on linux

**Comment by vgturtle127** [comment]
@cxd4 Well, regardless, you should git dat internets breh. BREH!

Anyway, has any progress been made in the way of fixing this? Just curious, since I want to start my testing as soon as possible, but I would like to have as many issues worked out as possible.

**Comment by JunielKatarn** [comment]
@vgturtle127 You can start testing. As explained in the thread, it's a case of mishandling unsupported hardware "addresses" (used at some point by the OOT ROM).
It only happens if you have the debugger enabled (which I always do).
Just disable the debugger when playing OOT and you should be fine.

**Comment by vgturtle127** [comment]
@JunielKatarn Well, I understand that, it's just I don't want to do a bunch of testing on a version of Project64 that is undergoing a large amount of cleanup, modernization, and being made more portable.

But at the same time I would like to start ASAP...

I don't know, what do you guys think?

**Comment by LegendOfDragoon** [comment]
@vgturtle127 there are certain things relevant to test that have little to do with the rewrite. For instance, Advanced Block Linking is buggy on some games. You can try testing games that have it on by default, and run sync cpu to see if it's stable. I'm not sure if it's on by default, but it's bugged in Perfect Dark and Nuclear Strike. If any of those 2 games have it on by default in the RDB, then you could make a PR to change the RDB settings.

As a reminder, sync cpu is very slow, so you want to use a very fast gfx plugin like Jabo 1.5.2 or something and HLE for audio as well. Be sure to enable fixed audio timing, when testing sync cpu.

**Comment by vgturtle127** [comment]
@LegendOfDragoon You keep talking about sync CPU, but I can't find it anywhere? Am I stupid? Where is it? Could you post a screenshot?

I'm sure I'm just missing it or something...

**Comment by LegendOfDragoon** [comment]
Enable debugger, then change CPU core style to Synchronize cores.

**Comment by project64** [comment]
@vgturtle127  you need the debugger on to see if like @LegendOfDragoon  explained above, it is more meant for me then for the general public which is why it is more hidden.

It will run the interpter and recompiler at the same time and at the end of each block of code it will make sure all the register match. So it might be slow, but it does mean if it is running like that then the recompiler is doing 100% as the interpter.

**Comment by vgturtle127** [comment]
@LegendOfDragoon @project64 OK, I found it. I thought you guys meant it was in the debugger settings. It's in the game's ROM settings. :laughing: 

Well, now that I feel stupid it's time to do some initial testing. :wink:

**Comment by vgturtle127** [comment]
Or maybe I won't do any initial testing?

![screenshot 5](https://cloud.githubusercontent.com/assets/3183676/12158579/c11e6af6-b4a0-11e5-850c-4da980ea64e3.png)

**Comment by LegendOfDragoon** [comment]
Before that "Break point found" message, did you get a message box saying "Sync Error"? If so, then that means that whatever settings you are using, are causing the recompiler to not function the same as the cpu interpreter. I guess ABL is enabled in that game in the RDB then. It should be off for that game, unless @project64 would rather fix the ABL issue(s), instead of accepting multiple pull requests to change the RDB settings.

Basically when there's a difference in the result between recompiler and interpreter, a "Sync Error" message box appears and once you hit ok, then that break point message appears.

**Comment by vgturtle127** [comment]
@LegendOfDragoon Yeah, I got the sync error.

Should I try Perfect Dark as well?

**Comment by LegendOfDragoon** [comment]
@vgturtle127 ok so that means the RDB could use some tweaking for Nuclear Strike. Yes, you can try Perfect Dark.

**Comment by vgturtle127** [comment]
@LegendOfDragoon Yeah, Perfect Dark does it too. Are there any other games you can think of that might be worth testing with sync CPU on?

**Comment by LegendOfDragoon** [comment]
@vgturtle127 Up to you really. It's just a matter of curiosity / interest. I basically only check, when I'm curious to see if a particular game I'm interested in is working as it should, or when I want to try tweaking recompiler settings in a game. For instance if I see a game using Protect Memory, I check to see if it actally needs it. Or if I run into a problem and suspect that the recompiler might be causing it, I sometimes check with sync cores.

**Comment by vgturtle127** [comment]
@LegendOfDragoon OK, do you know what RDB options are related to the sync CPU option? Like are there any consistent RDB options that all games that don't work or do work with sync CPU have?

**Comment by Frank-74** [comment]
Ocarina Of Time v1.2 (U), I get 'Unknown memory action' error.  I don't have the debugger enabled.

Turning of register caching fixes it.

**Comment by project64** [comment]
is this still an issue, or has it been fixed with the latest changes?

**Comment by LegendOfDragoon** [comment]
@JunielKatarn is this still an issue?

**Comment by JunielKatarn** [comment]
@LegendOfDragoon I'll confirm tonight.

**Comment by cxd4** [comment]
... and 26 hours later ...

**Comment by JunielKatarn** [comment]
The issue is resolved :)
